src/pool.rs

```diff
@@ -230,22 +230,27 @@ impl<T> Pool<T> {
     /// Restores all the garbage elements back to real elements. Effectively
     /// this is a rollback of all the remove operations that happened
     pub fn restore_garbage(&mut self) {
+        let mut carry_over = Vec::with_capacity(self.garbage.len());
         for idx in self.garbage.drain(..) {
             let mut item = PoolItem::Empty;
             std::mem::swap(&mut self.vec[idx], &mut item);
             self.vec[idx] = match item {
                 PoolItem::Filled(item) => PoolItem::Filled(item),
                 PoolItem::Tombstone(item) => PoolItem::Filled(item),
-                PoolItem::Empty => PoolItem::Empty,
+                PoolItem::Empty => {
```

Comment:
Is it even possible for the garbage element to be a `PoolItem::Filled`?

Looks like the only legal way for the element to become garbage is via the `Pool::remove` or `Pool::tombstone`.

Shouldn't we throw an error on `PoolItem::Filled` case?